internal/abci/abci.go

```diff
@@ -164,6 +172,29 @@ func NewAbciApp(ctx context.Context, cfg *AbciConfig, snapshotter SnapshotModule
 	if height == -1 {
 		height = 0 // negative means first start (no state table yet), but need non-negative for below logic
 	}
+
+	if bytes.Equal(appHash, []byte{0x42}) {
+		// This is an interesting corner case where the node crashes after the commit of block execution
+		// state but before the commit of the app hash. The chain state will have the app hash as "42".
+		// CometBFT seems to persist the FinalizeBlock responses, so during restart it calls "Commit"
+		// directly without calling FinalizeBlock and assumes that the application has the correct state
+		// persisted, whereas the application has the app hash as "42". This will cause the current block
+		// to be committed with an invalid app hash (sentry node blocksyncing will fail at this point).
+		// To fix this, we will store the last commit info in a file and use that as the application's
+		// app hash whenever we find the chain state to be inconsistent signified by the app hash as "42".
+		lc, err := loadLastCommitInfo(app.lastCommitInfoFileName)
+		if err != nil {
+			return nil, err
+		}
+
+		if height != lc.Height {
+			return nil, fmt.Errorf("height mismatch between chain state and last commit info: %d != %d", height, lc.Height)
+		}
+
+		appHash = lc.AppHash
+		log.Info("Recovered last commit info", zap.Int64("height", height), zap.String("appHash", hex.EncodeToString(appHash)))
```

Comment:
This should be rare, and a sign of crash, so Warn would be appropriate and helpful.

Also, in log statements, it would be best to be in the habbit of using our generic helpers like `log.Int` instead of the zap ones.  We want to minimally couple our packages with zap and leave that detail to the our `log` package